controllers/awscluster_controller_unit_test.go

```diff
@@ -0,0 +1,632 @@
+				{
+					APIVersion: clusterv1.GroupVersion.String(),
+					Kind:       "Cluster",
+					Name:       "capi-test-cluster",
+					UID:        "1",
+				},
+			}}},
+			ownerCluster: &clusterv1.Cluster{ObjectMeta: metav1.ObjectMeta{Name: "capi-test-cluster", Namespace: "default"}},
+			expectError:  false,
+		},
+		{
+			name:        "Should Reconcile successfully if no AWSCluster found",
+			expectError: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			g := NewWithT(t)
+			ctx := context.Background()
+			reconciler := &AWSClusterReconciler{
+				Client: testEnv.Client,
+			}
+			if tc.ownerCluster != nil {
+				g.Expect(testEnv.Create(ctx, tc.ownerCluster)).To(Succeed())
+				defer func(do ...client.Object) {
+					g.Expect(testEnv.Cleanup(ctx, do...)).To(Succeed())
+				}(tc.ownerCluster)
+			}
+			if tc.awsCluster != nil {
+				tc.awsCluster.Default()
+				g.Expect(testEnv.Create(ctx, tc.awsCluster)).To(Succeed())
+				defer func(do ...client.Object) {
+					g.Expect(testEnv.Cleanup(ctx, do...)).To(Succeed())
+				}(tc.awsCluster)
+				_, err := reconciler.Reconcile(ctx, ctrl.Request{
+					NamespacedName: client.ObjectKey{
+						Namespace: tc.awsCluster.Namespace,
+						Name:      tc.awsCluster.Name,
+					},
+				})
+				if tc.expectError {
+					g.Expect(err).ToNot(BeNil())
+				} else {
+					g.Expect(err).To(BeNil())
+				}
+			} else {
+				_, err := reconciler.Reconcile(ctx, ctrl.Request{
+					NamespacedName: client.ObjectKey{
+						Namespace: "default",
+						Name:      "test",
+					},
+				})
+				g.Expect(err).To(BeNil())
+			}
+		})
+	}
+}
+
+func TestAWSClusterReconcileOperations(t *testing.T) {
+	var (
+		reconciler AWSClusterReconciler
+		mockCtrl   *gomock.Controller
+		ec2Svc     *mock_services.MockEC2MachineInterface
+		elbSvc     *mock_services.MockELBInterface
+		networkSvc *mock_services.MockNetworkInterface
+		sgSvc      *mock_services.MockSecurityGroupInterface
+		recorder   *record.FakeRecorder
+	)
+
+	setup := func(t *testing.T, g *WithT, awsCluster *infrav1.AWSCluster) client.WithWatch {
+		t.Helper()
+		secret := &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-secret",
+				Namespace: "capa-system",
+			},
+			Data: map[string][]byte{
+				"AccessKeyID":     []byte("access-key-id"),
+				"SecretAccessKey": []byte("secret-access-key"),
+				"SessionToken":    []byte("session-token"),
+			},
+		}
+		csClient := fake.NewClientBuilder().WithObjects(awsCluster, secret).Build()
+
+		mockCtrl = gomock.NewController(t)
+		ec2Svc = mock_services.NewMockEC2MachineInterface(mockCtrl)
+		elbSvc = mock_services.NewMockELBInterface(mockCtrl)
+		networkSvc = mock_services.NewMockNetworkInterface(mockCtrl)
+		sgSvc = mock_services.NewMockSecurityGroupInterface(mockCtrl)
+
+		recorder = record.NewFakeRecorder(2)
```

Comment:
use `32` instead of `2` in NewfakeRecorder

Reply:
any specific reason for this?

Reply:
I have seen in cluster API `32` is mainly used as a buffer size for NewFakeRecorder(). If you are sure `2` will work then you can leave it

Reply:
atleast in CAPA I see it mostly as 2, so let's leave it for now, we can always increase it later on need basis.